Explore: Fix Browser title not updated on Navigation to Explore

axelavargas opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
When navigating to Explore page the document.title was not being updated, this was caused because Explore page is not using `Page` component that takes care of this functionality.

##Notes about the Approach
I decided to add the functionality on Explore page by duplicating the same logic we applied on `Page` component. That component is being used on other pages (Configuration, Alerting, etc...) but for the explore page will cause some unwanted styling issues and refactoring.

![ExploreDocumentTitle](https://user-images.githubusercontent.com/239999/119645384-43292b80-be1e-11eb-8999-0556de20ae6c.gif)


**Which issue(s) this PR fixes**:

<!--
- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #32898

**Special notes for your reviewer**:

axelavargas committed:
Explore: fix Browser title not updated on Navigation to Explore
- Update document.title on Explore page
- Add unit test and snapshot for Wrapper component

hugohaggmark commented on public/app/features/explore/Wrapper.tsx:
```diff
@@ -6,13 +6,32 @@ import { lastSavedUrl, resetExploreAction, richHistoryUpdatedAction } from './st
+import { Branding } from '../../core/components/Branding/Branding';
 
-interface WrapperProps extends GrafanaRouteComponentProps<{}, ExploreQueryParams> {
+// Libraries
+import { getNavModel, getTitleFromNavModel } from '../../core/selectors/navModel';
+
+// Types
+import { StoreState } from 'app/types';
```
Please don't add import comments, that's something we're trying to remove. Thank you.

axelavargas replied:
Fixed :) https://github.com/grafana/grafana/pull/34651/commits/0e6990946b6134881f68061c97c19d295ff8976d#diff-06aad0593d46c656f9c92b1a6e517f30e9758c1f1eff827194d81389ee5c74f6R12

hugohaggmark commented on public/app/features/explore/Wrapper.tsx:
```diff
@@ -46,9 +66,15 @@ export class Wrapper extends Component<WrapperProps> {
 };
 
-export default connect(null, mapDispatchToProps)(Wrapper);
+export default connect(mapStateToProps, mapDispatchToProps)(Wrapper);
```
This is an old way of setting up Redux connections refer to [this for the preferred way](https://github.com/grafana/grafana/blob/main/contribute/style-guides/redux.md#typing-of-connected-props).

axelavargas replied:
Thank you @hugohaggmark 🙌🏾  I will take a look :)

axelavargas replied:
I did the changes here as well https://github.com/grafana/grafana/pull/34651/commits/0e6990946b6134881f68061c97c19d295ff8976d#diff-06aad0593d46c656f9c92b1a6e517f30e9758c1f1eff827194d81389ee5c74f6R15

hugohaggmark reviewed (commented):
Code looks great, but I'm not sure about the title part. I think `Explore - Grafana` should be enough. `Explore: Explore - Grafana` looks weird to me.

axelavargas committed:
Explore: refactor Wrapper component to use new connected redux props approach

- Remove unnecesary comments
- Wrapper component is now inferring connected props
- Update snapshot

axelavargas commented:
> Code looks great, but I'm not sure about the title part. I think `Explore - Grafana` should be enough. `Explore: Explore - Grafana` looks weird to me.

Thanks, @hugohaggmark for the review, indeed it looks weird, I did the changes here https://github.com/grafana/grafana/pull/34651/commits/0e6990946b6134881f68061c97c19d295ff8976d#diff-06aad0593d46c656f9c92b1a6e517f30e9758c1f1eff827194d81389ee5c74f6R35, I had to use directly the value from `${navModel.main.text}`, my old approach was using `getTitleFromNavModel(navModel);` but unfortunately this doesn't work in `Explore` page because the `main.text` and the `node.text` are the same text, and this is something generated in the backend.

hugohaggmark commented:
> > Code looks great, but I'm not sure about the title part. I think `Explore - Grafana` should be enough. `Explore: Explore - Grafana` looks weird to me.
> 
> Thanks, @hugohaggmark for the review, indeed it looks weird, I did the changes here [0e69909#diff-06aad0593d46c656f9c92b1a6e517f30e9758c1f1eff827194d81389ee5c74f6R35](https://github.com/grafana/grafana/commit/0e6990946b6134881f68061c97c19d295ff8976d#diff-06aad0593d46c656f9c92b1a6e517f30e9758c1f1eff827194d81389ee5c74f6R35), I had to use directly the value from `${navModel.main.text}`, my old approach was using `getTitleFromNavModel(navModel);` but unfortunately this doesn't work in `Explore` page because the `main.text` and the `node.text` are the same text, and this is something generated in the backend.
> 
> _ps. I updated the gif in this PR with the new title_

Outstanding, will do another pass now.

hugohaggmark commented on public/app/features/explore/Wrapper.tsx:
```diff
@@ -1,18 +1,43 @@
-import React, { Component } from 'react';
-import { connect } from 'react-redux';
+import React, { PureComponent } from 'react';
```
The change from Component to PureComponent should be fine AFAIK but maybe @ivanahuckova or @aocenas can confirm this too?

ivanahuckova replied:
As far as I know, the change should be okay. 🙂

aocenas replied:
I don't think it should break anything but I have a feeling:
```
navModel: getNavModel(state.navIndex, 'explore'),
```
will break any benefit as it seems to create a new object every time.

axelavargas replied:
great 🙌🏾, thank you @ivanahuckova for taking a look, I will proceed to merge :)

axelavargas replied:
> I don't think it should break anything but I have a feeling:
> 
> ```
> navModel: getNavModel(state.navIndex, 'explore'),
> ```
> 
> will break any benefit as it seems to create a new object every time.

Hey, @aocenas thanks for checking out this PR 😄 , I was looking into `getNavModel` function, even though it generates a new object every time, the values are always the same if the parameters don't change, and because PureComponent only does a shallow comparison on the `ComponentShouldUpdate` I think will keep the benefits. I also verified the children's `Wrapper` component to confirm they are also PureComponents only to make sure are not breaking anything.

let me know if I am mistaken(I still lacking context in the project 😅 ) , I will merge the PR but ofc, if we find later on this, don't add any benefits we can put it back as before :)

hugohaggmark reviewed (approved):
Looks outstanding! 🎉  

The change from Component to PureComponent should be fine AFAIK but there might be some edge case that I'm unaware of.

axelavargas committed:
Merge branch 'main' into axelav/explore-title-not-updated

axelavargas committed:
Explore: Update snapshot Wrapper test

ivanahuckova reviewed (approved):
Thank you for working on this @axelavargas! 🧡 It has been in the backlog for quite some time and I am very happy that Explore has finally title.✨ LGTM!
